electrolytes/solvation_shell_utils.py

```diff
@@ -3,44 +3,70 @@
 import numpy as np
 
 
-def rmsd(a, b):
+def rmse(a, b):
     """
-    Compute the root mean squared distance between two sets of coordinates
+    Compute the root mean squared error between two sets of pairwise displacements
     Args:
-        a: numpy array of positions, shape [N_atoms, 3]
-        b: numpy array of positions, shape [N_atoms, 3]
+        a: numpy array of pairwise displacements, shape [N_atoms, N_atoms, 3]
+        b: numpy array of pairwise displacements, shape [N_atoms, N_atoms, 3]
     """
-    return np.sqrt(np.mean(np.linalg.norm(b - a, axis=1)))
+    return np.sqrt(np.mean(np.sum((b - a) ** 2, axis=2)))
 
 
-def filter_by_rmsd(coords, n=20):
+def filter_by_rmse(coords, n=20):
     """
-    From a set of coordinates, determine the n most diverse, where "most diverse" means "most different, in terms of RMSD"
-
-    Because of the way that IonSolvR data is organized, should be able to just directly calculate rmsd on numpy coordinate arrays
+    From a set of coordinates, determine the n most diverse, where "most diverse" means "most different, in terms of minimum RMSE.
+    We operate on pairwise distances so that the function is invariant to translation and rotation.
+    Note: The Max-Min Diversity Problem is in general NP-hard. This algorithm generates a candidate solution to MMDP for these coords
+    by assuming that the point 0 is actually in the MMDP set (which there's no reason a priori to assume). As a result, if we shuffled the order of coords, we would likely get a different result.
 
     Args:
         coords: list of np.ndarrays of atom coordinates. Must all have the same shape ([N_atoms, 3]), and must all reflect the same atom order!
             Note that this latter requirement shouldn't be a problem, specifically when dealing with IonSolvR data.
         n: number of most diverse coordinates to return
     """
-
+    pairwise_disps = [
+        coord[np.newaxis, :, :] - coord[:, np.newaxis, :] for coord in coords
+    ]
     states = {0}
-    min_rmsds = np.array([rmsd(coords[0], coords[i]) for i in range(len(coords))])
-    for i in range(n - 1):
+    min_rmsds = np.array(
+        [rmse(pairwise_disps[0], pairwise_disp) for pairwise_disp in pairwise_disps]
+    )
+    for _ in range(n - 1):
         best = np.argmax(min_rmsds)
         min_rmsds = np.minimum(
             min_rmsds,
-            np.array([rmsd(coords[best], coords[i]) for i in range(len(coords))]),
+            np.array(
+                [
+                    rmse(pairwise_disps[best], pairwise_disp)
+                    for pairwise_disp in pairwise_disps
+                ]
+            ),
         )
         states.add(best)
 
     return [coords[i] for i in states]
 
 
+def wrap_positions(positions, lattices):
```

Comment:
Sorry, I'm probably just being thick here. I assume this function is supposed to fix molecules if they spanned PBC, but won't it break up molecules if the cluster is larger than half a box size in any dimension? I can see that not being too common, but just want to check.

Reply:
I think you're right, but I'm not sure how we would deal with that situation, since there would be no way to keep everything within a ball of diameter `lattices /2`. For the low concentration runs where we are only taking a single solvation shell, I don't think this would ever be the case, but it might become an issue for the higher concentration runs. In that case, would it make sense to operate on the original, unwrapped trajectories from the MD simulation? Or is there a way to deal with this during the DFT single point calculations?

Reply:
1. Presumably we'll want a size cutoff on clusters/networks, so I'm not terribly worried about massive, sprawling clusters
2. So, you could look _ex post facto_ look for residues (you know which atoms are associated with which residue from MDAnalysis) with broken bonds. More computationally expensive, since you either need to look at pairwise distances or generate a graph. But whatever; once you know which residues are broken, you can just shift the atoms in such a way that a) the graph is no longer broken and b) the residue as a whole is less than some distance away from the rest of the group. Not nearly as clean as this, but we can catch this edge case if we really want to.

Reply:
Gotcha. @levineds thoughts?

Reply:
The key feature is that molecules/closely interacting systems don't get broken up. After a contraction procedure (or a wrapping procedure as you call it), there's not really a problem if there are atoms that are outside the lattice box. That is, at the end, we can let things span the PBC boundary by going outside the box. There are some known algorithms for doing this. So I agree we can just leave this for now and it probably won't be an issue for smaller clusters, but as long as we store the positions and lattice parameters somewhere convenient, we can post-process to ensure that systems are not broken up by PBCs.